src/RecurringAllowanceManager.sol

```diff
@@ -0,0 +1,306 @@
+    ///
+    /// @param recurringAllowance Details of the recurring allowance.
+    function revoke(RecurringAllowance calldata recurringAllowance)
+        external
+        requireSender(recurringAllowance.account)
+    {
+        bytes32 hash = getHash(recurringAllowance);
+        _isRevoked[hash][recurringAllowance.account] = true;
+        emit RecurringAllowanceRevoked(hash, recurringAllowance.account, recurringAllowance);
+    }
+
+    /// @notice Withdraw tokens using a recurring allowance.
+    ///
+    /// @param recurringAllowance Details of the recurring allowance.
+    /// @param value Amount of token attempting to withdraw (wei).
+    function withdraw(RecurringAllowance calldata recurringAllowance, uint160 value)
+        external
+        requireSender(recurringAllowance.spender)
+    {
+        // early return if no value spent
+        if (value == 0) return;
+
+        // require recurring allowance is approved and not revoked
+        if (!isAuthorized(recurringAllowance)) revert Unauthorized();
+
+        CycleUsage memory currentCycle = getCurrentCycle(recurringAllowance);
+        uint256 totalSpend = value + currentCycle.spend;
+
+        // check total spend value does not overflow max value
+        if (totalSpend > type(uint160).max) revert WithdrawValueOverflow(totalSpend);
+
+        // check total spend value does not exceed recurring allowance
+        if (totalSpend > recurringAllowance.allowance) {
+            revert ExceededRecurringAllowance(totalSpend, recurringAllowance.allowance);
+        }
+
+        bytes32 hash = getHash(recurringAllowance);
+
+        // save new withdraw for active cycle
+        currentCycle.spend = uint160(totalSpend);
+        _lastUpdatedCycle[hash][recurringAllowance.account] = currentCycle;
+        emit RecurringAllowanceWithdrawn(
+            hash,
+            recurringAllowance.account,
+            recurringAllowance.token,
+            CycleUsage(currentCycle.start, currentCycle.end, uint160(value))
+        );
+
+        // call account to transfer tokens
+        _withdraw({
+            owner: recurringAllowance.account,
+            recipient: recurringAllowance.spender,
+            token: recurringAllowance.token,
+            value: value
+        });
+    }
+
+    /// @notice Hash a RecurringAllowance struct for signing.
+    ///
+    /// @dev Prevent phishing permits by making the hash incompatible with EIP-191/712.
+    /// @dev Include chainId and contract address in hash for cross-chain and cross-contract replay protection.
+    ///
+    /// @param recurringAllowance Details of the recurring allowance.
+    ///
+    /// @return hash Hash of the recurring allowance and replay protection parameters.
+    function getHash(RecurringAllowance memory recurringAllowance) public view returns (bytes32) {
+        return keccak256(abi.encode(recurringAllowance, block.chainid, address(this)));
```

Comment:
might be nice to do this as a a 712 so that we have a domain hash that includes chain ID, address(this) following that convention

Reply:
e.g. https://github.com/base-org/scw-recovery-signer/blob/main/src/RecoverySigner.sol#L69

Reply:
I was thinking we'd want this to actually prevent 191/712 compatibility for phishing concerns, similar to our permission signing from before? Maybe now that the raw parameters are more legible, we can actually do 712 safely, but this was the intuition.

I could also see a world where our wallet's `signTypedData` RPC redirects to our permission approval screen if it detects the domain+typehash match this contract.